File: src/piwardrive/route_optimizer.py
"""Route optimization helpers."""

from __future__ import annotations

import math
from typing import Iterable, Tuple, List

Coord = Tuple[float, float]


def suggest_route(
    points: Iterable[Coord],
    *,
    cell_size: float = 0.001,
    steps: int = 5,
    search_radius: int = 5,
) -> List[Coord]:
    """Return waypoints that maximize coverage based on ``points``.

    Parameters
    ----------
    points:
        Sequence of ``(lat, lon)`` coordinates in chronological order.
    cell_size:
        Grid resolution in degrees used to mark visited areas.
    steps:
        Number of waypoints to return.
    search_radius:
        How far to search for new cells around the current location, in grid
        cells.
    """
    pts = [(float(lat), float(lon)) for lat, lon in points]
    if not pts:
        return []

    def to_cell(lat: float, lon: float) -> tuple[int, int]:
        return math.floor(lat / cell_size), math.floor(lon / cell_size)

    visited = {to_cell(lat, lon) for lat, lon in pts}
    cur_cell = to_cell(*pts[-1])

    route: List[Coord] = []
    for _ in range(steps):
        best: tuple[int, int] | None = None
        best_dist: int | None = None
        for dx in range(-search_radius, search_radius + 1):
            for dy in range(-search_radius, search_radius + 1):
                cell = (cur_cell[0] + dx, cur_cell[1] + dy)
                if cell in visited:
                    continue
                dist = dx * dx + dy * dy
                if best_dist is None or dist < best_dist:
                    best = cell
                    best_dist = dist
        if best is None:
            break
        visited.add(best)
        cur_cell = best
        lat = (best[0] + 0.5) * cell_size
        lon = (best[1] + 0.5) * cell_size
        route.append((lat, lon))
    return route


__all__ = ["suggest_route"]

File: webui/src/routeOptimizer.js
export function suggestRoute(
  points,
  cellSize = 0.001,
  steps = 5,
  searchRadius = 5
) {
  const pts = points.map((p) => [Number(p[0]), Number(p[1])]);
  if (pts.length === 0) return [];
  const toCell = (lat, lon) => [
    Math.floor(lat / cellSize),
    Math.floor(lon / cellSize),
  ];
  const visited = new Set(pts.map((p) => toCell(p[0], p[1]).join(',')));
  let cur = toCell(pts[pts.length - 1][0], pts[pts.length - 1][1]);
  const route = [];
  for (let i = 0; i < steps; i++) {
    let best = null;
    let bestDist = null;
    for (let dx = -searchRadius; dx <= searchRadius; dx++) {
      for (let dy = -searchRadius; dy <= searchRadius; dy++) {
        const cell = [cur[0] + dx, cur[1] + dy];
        const key = cell.join(',');
        if (visited.has(key)) continue;
        const dist = dx * dx + dy * dy;
        if (bestDist === null || dist < bestDist) {
          best = cell;
          bestDist = dist;
        }
      }
    }
    if (!best) break;
    visited.add(best.join(','));
    cur = best;
    const lat = (best[0] + 0.5) * cellSize;
    const lon = (best[1] + 0.5) * cellSize;
    route.push([lat, lon]);
  }
  return route;
}
